Yosemite/Yosemite/Tools/Stats/LeaderboardStatsConverter.swift

```diff
@@ -0,0 +1,82 @@
+import Foundation
+import class Aztec.HTMLParser
+import struct Aztec.Element
+
+/// Helper functions to  convert a `Leaderboard` type into a `TopEearnerStat`
+///
+struct LeaderboardStatsConverter {
+
+    /// Returns the ids of the products associated with a top products leaderboar
+    ///
+    static func topProductsIDs(from topProducts: Leaderboard) -> [Int64] {
```

Comment:
I use static functions when `self` is not a dependency on the work that I have to do.
I think that using a static function expresses the reader a clear intention that:
- The function does not depend on any stored state
- The function does not mutate any state within the object.

Reply:
in my previous work, we also had the same reasons for using `static`. personally, I slightly prefer using member functions for all in case a state is added and it looks nicer without `Self` 😅  I'm fine with either way though, I remember having a conversation with @shiki earlier just thought Jayson might be interested 🙂

Reply:
I'm happy to follow any convention that the team has previously decided tho! :-)

Reply:
I want to say that it's fine to have `static` methods on helpers. For the case of `LeaderboardStatsConverter`, I think it's grown beyond a simple helper. On a quick glance, the methods do not look related to me at all. First, the comment is:

```swift
/// Helper functions to  convert a `Leaderboard` type into a `TopEearnerStat`
```

But only one of the 4 methods return `TopEearnerStatItem`. The _helper_ failed in the contract that it tried to represent, which is converting something to `TopEearnerStat`. 

The separation of the 3 additional _accessible_ methods also adds friction to understanding how the methods should be used together. And it will get cumbersome to refactor since they are all accessible and you don't know where they are truly used. 

Based on how I see them being used in [`loadProduct`](https://github.com/woocommerce/woocommerce-ios/blob/9a0409d302549e9548744ef8c138f739301e284f/Yosemite/Yosemite/Stores/StatsStoreV4.swift#L337-L340), a possible improvement would be using the  `*UseCase` ([example](https://github.com/woocommerce/woocommerce-ios/blob/develop/Yosemite/Yosemite/Stores/ProductReview/RetrieveProductReviewFromNoteUseCase.swift)) strategy. We can gather these three lines maybe:

```swift
        // Workout if we have stored all products that relate to the given leaderboard
        let topProductIDs = LeaderboardStatsConverter.topProductsIDs(from: topProducts)
        let topStoredProducts = loadStoredProducts(siteID: siteID, productIDs: topProductIDs)
        let missingProductsIDs = LeaderboardStatsConverter.missingProductsIDs(from: topProducts, in: topStoredProducts)
```

And then let's ask ourselves. What are the inputs and outputs? For the lines above, I think the input is `Leaderboard` and the output is `[Int64]`. Notice that we don't need `LeaderboardStatsConverter.topProductsIDs` for this routine so we can **eliminate** that right away. The `*UseCase` can probably be something like this:

```swift
struct DoSomethingUseCase {
    private let leaderboard: Leaderboard

    func getMissingProductIds() -> [Int64] {
        // call topProductIds()
        // call loadStoredProducts
        return missingProductIds
    }

    private func topProductIds(from: Leaderboard) -> [Int64] {
    
    }
}
```

This `UseCase` is still testable in isolation. The big win here for me is that there is only 1 method and everything else is private. It is much easier to reason about how to use this kind of _helper_. In the end, you can think of the `UseCase` as a `static func` that contains **a lot** of `private static func`. 😄 

There's a lot that goes into these of course so I definitely skipped a lot of things to discuss here. 

----

And as I'm writing this, I realized I didn't answer Jaclyn's points. 😄 

> I slightly prefer using member functions for all in case a state is added 

I agree with Ernesto about having clearer intentions that there is no state. Though this comes with the problem of having too many arguments. I think there needs to be a balance. With the `UseCase` strategy, we've eliminated both problems. 

---

Another point! 😅 

Our Yosemite Store classes tend to get really large. Eventually, we're really going to have to split things up into logical chunks. It would be a shame if we split them using multiple `static funcs` instead of business-logic UseCases: 

https://github.com/woocommerce/woocommerce-ios/blob/9a0409d302549e9548744ef8c138f739301e284f/Yosemite/Yosemite/Stores/ProductReviewStore.swift#L104-L114

---

In summary:

- `static func`: Good for _short_ helpers but probably not as the number lines get longer. And probably not for business logic.
- private members: Good for short classes but as the class gets bigger, the number of methods grow and everything becomes harder to digest. 
- logical chunks (like UseCase): good for encapsulating a single input and output that involves a lot of different routines. 

Note: I'm not saying what's the _best_ thing here. I'm just sharing what I think when to use what. 🙂 And every day my opinions change! 😁